### PR TITLE
Implement find builtin command

### PR DESCRIPTION
**Pull request by matt-dz**
## Summary
- Add a sandboxed `find` builtin for searching directory trees with expression-based predicates
- Supports `-name`, `-iname`, `-path`, `-ipath`, `-type`, `-size`, `-empty`, `-newer`, `-mtime`, `-mmin`, `-maxdepth`, `-mindepth`, `-print`, `-print0`, `-prune`, `-true`, `-false`, and logical operators (`!`, `-a`, `-o`, `()`)
- Blocks unsafe predicates (`-exec`, `-delete`, `-regex`, `-fprint`, etc.) for sandbox safety
- Uses recursive-descent expression parser with bounded AST (max 64 depth, 256 nodes) and explicit-stack DFS traversal (capped at 256 depth)

## Implementation
- `interp/builtins/find/find.go` — Command handler, arg parsing, DFS walker
- `interp/builtins/find/expr.go` — AST types + recursive-descent parser
- `interp/builtins/find/eval.go` — Expression evaluator
- `interp/builtins/find/match.go` — Glob/type/size matching helpers
- 24 YAML test scenarios covering basic, predicates, logic, depth, size, output, prune, errors, and sandbox categories
- Import allowlist updated with `path.Match` and `strings.ToLower`

## Test plan
- [x] `go test ./interp/... ./tests/...` — all pass
- [x] `RSHELL_BASH_TEST=1 go test ./tests/ -run TestShellScenariosAgainstBash` — all pass (ordering-sensitive tests marked `skip_assert_against_bash` due to ReadDir returning sorted entries vs GNU find's filesystem order)
- [x] Pentest scenarios for resource exhaustion, symlink loops, blocked predicates
- [x] Code review cycle

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: interp/builtins/find/expr.go [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+		numStr = numStr[1:]
+	} else if strings.HasPrefix(numStr, "-") {
+		cmp = -1
+		numStr = numStr[1:]
+	}
+	n, err := strconv.Atoi(numStr)
+	if err != nil {
+		return nil, fmt.Errorf("find: invalid argument '%s' to %s", val, kindName(kind))
+	}
+	return &expr{kind: kind, numVal: int64(n), numCmp: cmp}, nil
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/builtins/find/expr.go]: [Claude Opus 4.6] Fixed — replaced `strconv.Atoi` with `strconv.ParseInt(numStr, 10, 64)` in both `parseNumericPredicate` and `parseSize`. Kept `Atoi` in `parseDepthOption` since depth values are small integers stored in `int` fields.

**Review by matt-dz — COMMENTED** [review]
## Review Summary — Round 6

**Reviewer:** Claude Opus 4.6 (self-review)
**Scope:** Full PR diff (find builtin implementation + round 5 fixes)
**Overall:** **Safe to merge** — no security issues found

### Security Assessment

The implementation is well-sandboxed:
- All filesystem access goes through `callCtx.StatFile`, `callCtx.LstatFile`, `callCtx.ReadDir`, `callCtx.OpenFile` — never directly via `os.*`
- Dangerous predicates (`-exec`, `-execdir`, `-delete`, `-ok`, `-okdir`, `-fls`, `-fprint*`, `-regex`, `-iregex`) are blocked at parse time
- AST complexity is bounded (`maxExprNodes=256`, `maxExprDepth=64`)
- Traversal depth is capped (`maxTraversalDepth=256`)
- Symlink loop detection uses file identity (dev+inode / volume serial+file index) with path-based fallback
- `-newer` reference paths go through the sandbox `StatFile` wrapper
- Windows `FileIdentity` uses `syscall.CreateFile` with `dwDesiredAccess=0` (metadata-only) on paths already validated by the sandbox

### Findings

| # | Priority | File | Finding |
|---|----------|------|---------|
| 1 | <sub><sub>P2 Badge</sub></sub> | `tests/scenarios/cmd/find/` | Missing scenario test for duplicate `-newer` dedup |
| 2 | <sub><sub>P3 Badge</sub></sub> | `interp/builtins/find/match.go:193` | Trailing `\` in glob pattern fails to match literal backslash |

### Positive Observations

- Blocked predicates are comprehensive and enforced early at parse time
- The recursive-descent parser elegantly solves the argument-stealing problem (e.g. `find . -name -maxdepth` correctly treats `-maxdepth` as the `-name` pattern)
- Overflow-safe ceiling division in `compareSize` is well-guarded
- `newerCache` + `newerErrors` pattern avoids redundant stat calls and provides consistent error behavior
- Context cancellation is checked at every loop iteration in the traversal
- AST depth and node limits prevent expression-based DoS
- Good test coverage with 50+ scenario files across basic, predicates, logic, depth, errors, sandbox, symlinks, and output categories

**Comment by matt-dz** [comment]
@codex review

File: interp/builtins/find/match.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+			case '[':
+				// Character class — delegate to path.Match for the class portion.
+				if nx < len(name) {
+					matched, width := matchClass(pattern[px:], name[nx])
+					if matched {
+						px += width
+						nx++
+						continue
+					}
+				}
```
Comment (matt-dz) [inline interp/builtins/find/match.go]: **<sub><sub>P1 Badge</sub></sub>  `pathGlobMatch` doesn't treat malformed brackets as literals — bash divergence in `-path`**

While `matchGlob` (used by `-name`) was correctly fixed to fall back to literal comparison when `path.Match` errors on malformed patterns like `[`, the `pathGlobMatch` function (used by `-path` and `-ipath`) was **not** fixed.

When `matchClass` returns `(false, 0)` for a malformed bracket, `pathGlobMatch` falls through to the star-backtrack check. Without a star active, it returns `false` — meaning `-path 'dir/[sub/file.txt'` will **never** match a file at that path.

**Proof:**
```bash
# GNU find (correct):
$ mkdir -p 'dir/[sub' && touch 'dir/[sub/file.txt'
$ find dir -path 'dir/[sub/file.txt'
dir/[sub/file.txt

# rshell (broken): returns nothing
```

**Remediation:** When `matchClass` returns `(false, 0)` indicating a malformed class, fall back to literal `[` comparison instead of failing the match:

```suggestion
			case '[':
				// Character class — delegate to matchClass for the class portion.
				if nx < len(name) {
					matched, width := matchClass(pattern[px:], name[nx])
					if matched {
						px += width
						nx++
						continue
					}
					// Malformed class (width==0) — treat '[' as literal.
					if width == 0 && pattern[px] == name[nx] {
						px++
						nx++
						continue
					}
				}
```

File: tests/scenarios/cmd/find/predicates/newer_dedup.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    - path: dir/a.txt
+      content: "a"
+      chmod: 0644
+input:
+  allowed_paths: ["$DIR"]
+  script: |+
+    find dir -newer nonexist -o -newer nonexist
+expect:
+  stdout: ""
+  stderr_contains: ["find: 'nonexist'"]
```
Comment (matt-dz) [inline tests/scenarios/cmd/find/predicates/newer_dedup.yaml]: <sub><sub>P2 Badge</sub></sub> **`stderr_contains` doesn't verify dedup — only checks presence, not uniqueness**

The test description says "duplicate -newer refs produce only one error line" but `stderr_contains: ["find: 'nonexist'"]` only checks that the string is present — it would still pass if the error appeared twice.

Since `2>&1 | wc -l` isn't available (no fd duplication support), use exact `stderr` match instead. The error message format is platform-dependent, but you could use `stderr_contains` paired with a Go-level test that counts lines, or restructure the test to capture exact output.

Alternatively, add a Go unit test in `find_test.go` that programmatically invokes the find builtin with duplicate `-newer` refs and asserts exactly one `Errf` call.

**Review by matt-dz — COMMENTED** [review]
## Review Summary — Round 10 (Self-Review)

Reviewed commit `4afc9fb` (round 9 changes) on top of the previously-reviewed `5d1151a`.

### Overall Assessment: **Safe to merge** ✓

**No findings.** The delta is 3 files / 20 lines:
1. Empty parentheses detection in `parseUnary` with GNU find-compatible error message — correct
2. Star + malformed bracket backtracking test cases — correct
3. `empty_parens.yaml` scenario test — correct

All prior findings from rounds 7–9 have been addressed. All review threads are resolved. 0 unresolved comments remain.

### Security: PASS
- No new filesystem access patterns
- No new imports
- Expression parser change is purely additive (early return on empty parens)

### Bash Compatibility: PASS
- Empty parens error message now matches GNU find format
- Scenario test validates the behavior

### Test Coverage: Adequate
- `empty_parens.yaml` covers the new code path
- Star + malformed bracket test covers the interaction

File: interp/builtins/find/match.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+// The shell normalises all paths to forward slashes on all platforms,
+// so hardcoding '/' is correct even on Windows.
+func baseName(p string) string {
+	// Strip trailing slashes.
+	for len(p) > 1 && p[len(p)-1] == '/' {
+		p = p[:len(p)-1]
+	}
+	for i := len(p) - 1; i >= 0; i-- {
+		if p[i] == '/' {
+			return p[i+1:]
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/builtins/find/match.go]: [Claude Opus 4.6] Fixed: `baseName("/")` now returns `"/"` instead of `""`. Added `TestBaseNameEdgeCases` unit test covering root path, trailing slashes, and normal paths.

File: interp/builtins/find/find.go [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+	i := 0
+
+	// Parse leading global options.
+	for i < len(args) {
+		if args[i] == "-L" {
+			followLinks = true
+			i++
+		} else if args[i] == "-P" {
+			// -P is default (no follow).
+			i++
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/builtins/find/find.go]: [Claude Opus 4.6] Fixed: `-P` now resets `followLinks = false`, so `-L -P` correctly disables symlink following (last option wins). Added `P_overrides_L.yaml` scenario test.

File: interp/builtins/find/find.go [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+		// Descend into directories unless pruned or beyond maxdepth.
+		if entry.info.IsDir() && !prune && entry.depth < maxDepth {
+			// With -L, check for symlink loops.
+			if useFileID {
+				if id, ok := callCtx.FileIdentity(entry.path, entry.info); ok {
+					if firstPath, seen := visitedID[id]; seen {
+						callCtx.Errf("find: File system loop detected; '%s' is part of the same file system loop as '%s'.\n",
+							entry.path, firstPath)
+						failed = true
+						continue
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/builtins/find/find.go]: [Claude Opus 4.6] Fixed: cycle detection now uses **ancestor-chain tracking** instead of a global visited set. Each stack entry carries its own `ancestorIDs` map (root→parent), so multiple symlinks to the same target directory are traversed correctly without false loop errors. Only actual ancestor cycles (directory is its own ancestor) are flagged.

Added `multiple_links_same_target.yaml` scenario test verifying two symlinks to the same target are both traversed with `-L`.

**Comment by matt-dz** [comment]
@codex make a comprehensive code and security reviews

File: interp/builtins/find/expr.go [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+	exprFalse                  // -false
+	exprAnd                    // expr -a expr  or  expr expr (implicit)
+	exprOr                     // expr -o expr
+	exprNot                    // ! expr  or  -not expr
+)
+
+// sizeUnit holds a parsed -size predicate value.
+type sizeUnit struct {
+	n    int64 // magnitude (always positive)
+	cmp  int   // -1 = less than, 0 = exact, +1 = greater than
```
Comment (matt-dz) [inline interp/builtins/find/expr.go]: <sub><sub>P2 Badge</sub></sub> **Magic int for comparison operator** (Effective Go: Names)

`cmp int // -1 = less than, 0 = exact, +1 = greater than` appears in `sizeUnit`, `expr.numCmp`, `compareNumeric`, `compareSize`, and all the eval functions. A named type with constants would be clearer and prevent accidental misuse:

```go
type cmpOp int

const (
	cmpLess  cmpOp = -1
	cmpExact cmpOp = 0
	cmpMore  cmpOp = 1
)
```

Then `compareNumeric(actual, target int64, cmp cmpOp)` — the type system prevents passing an arbitrary int.

File: interp/builtins/find/find.go [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+// Cmd is the find builtin command descriptor.
+var Cmd = builtins.Command{Name: "find", MakeFlags: builtins.NoFlags(run)}
+
+func run(ctx context.Context, callCtx *builtins.CallContext, args []string) builtins.Result {
+	// Parse global options (-L) and separate paths from expression.
+	followLinks := false
+	i := 0
+
+	// Parse leading global options.
+	for i < len(args) {
```
Comment (matt-dz) [inline interp/builtins/find/find.go]: <sub><sub>P2 Badge</sub></sub> **`if-else-if` chain should be a `switch`** (Effective Go: Control Structures/Switch)

Effective Go recommends using `switch` instead of `if-else-if` chains for multi-way conditionals. This chain comparing `args[i]` against string literals is a textbook case:

```suggestion
	for i < len(args) {
		switch args[i] {
		case "-L":
			followLinks = true
			i++
		case "-P":
			followLinks = false
			i++
		case "-H":
			callCtx.Errf("find: -H is not supported\n")
			return builtins.Result{Code: 1}
		case "--":
			i++
			goto pathParsing
		default:
			goto pathParsing
		}
	}
```

(The `goto` is just one approach — a labeled break or extracting a helper also works.)

File: interp/allowed_paths_internal_test.go [hunk 9]
```diff
@@ -96,8 +96,8 @@ func TestAllowedPathsExecNonexistent(t *testing.T) {
 
 func TestAllowedPathsExecViaPathLookup(t *testing.T) {
 	dir := t.TempDir()
-	// "find" is resolved via PATH (not absolute), but /bin and /usr are not allowed
-	_, stderr, exitCode := runScriptInternal(t, `find`, dir,
+	// "sed" is resolved via PATH (not absolute), but /bin and /usr are not allowed
+	_, stderr, exitCode := runScriptInternal(t, `sed`, dir,
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/allowed_paths_internal_test.go]: [Claude Opus 4.6] Fixed in 3d540c5 — the test now uses `date` (not a builtin) and constructs the runner directly without `runScriptInternal`'s exec handler override, so the default `noExecHandler` properly blocks the command. No more guessing games with the builtin list.

File: interp/builtins/find/match.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+					}
+				}
+			case '\\':
+				// Escape: next character is literal.
+				px++
+				if px >= len(pattern) {
+					// Trailing backslash — treat as literal '\\'.
+					if nx < len(name) && name[nx] == '\\' {
+						nx++
+						continue
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/builtins/find/match.go]: [Claude Opus 4.6] Verified against GNU find on `debian:bookworm-slim` — `find . -name "\\"` (trailing backslash) DOES match a file named `\`. Our behavior already matches bash. No change needed.

**Comment by matt-dz** [comment]
@codex make a comprehensive code and security reviews

**Comment by matt-dz** [comment]
@codex make a comprehensive code and security reviews

**Comment by matt-dz** [comment]
Hey @thieman — addressed your `*os.File` feedback (narrowed to `fs.ReadDirFile`) and picked up the remaining bot comments:

- **Context cancellation**: `find` now returns exit 1 when cancelled mid-traversal instead of silently succeeding with partial output.
- **Future-dated file mtime**: Removed the negative delta clamp so `-mtime 0` no longer incorrectly matches future-dated files, matching GNU find.
- **Symlink cycle fallback** (path-based detection): Acknowledged the limitation but we think the tradeoff is acceptable here — `FileIdentity` works on all supported platforms, and the `maxTraversalDepth=256` hard cap provides defense-in-depth for the rare case where it doesn't. Happy to revisit if you feel differently.

Ready for another look when you have a moment!

**Comment by matt-dz** [comment]
@chatgpt-codex-connector — addressed your latest comment:

- **Dangling -newer refs under -L** (f4b613a): Both eager validation and `evalNewer` now fall back to `LstatFile` when `StatFile` fails with `ErrNotExist` under `-L`, matching GNU find. Added scenario test `newer_dangling_symlink_L`.

Ready for another look!
